File: p524m/find_longest_word.cpp
// Build program by
// g++ -std=c++11 find_longest_word.cpp -o test && ./test && rm -rf test

#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string findLongestWord(string s, vector<string>& d)
    {
        string ans = "";
        for (string& w : d) {
            if (w.length() > ans.length() || (w.length() == ans.length() && w < ans)) {
                int i = 0;
                for (char c : s) {
                    if (i < w.length() && c == w[i])
                        i++;
                }
                if (i == w.length())
                    ans = w;
            }
        }
        return ans;
    }
};
// TEST
struct Test {
    string s;
    vector<string> d;
    string expected;
    void run()
    {
        Solution sol;
        string actual = sol.findLongestWord(s, d);
        cout << "The longest word in [" << endl;
        for (const string& w : d) {
            cout << w << ", ";
        }
        cout << "\n] matching \"" << s << "\" -> " << actual << endl;
        assert(actual == expected);
    }
};
int main(int argc, char* argv[])
{
    vector<Test> tests = {
        { "abpcplea", { "ale", "apple", "monkey", "plea" }, "apple" },
        { "abpcplea", { "a", "b", "c", "d" }, "a" },
        { "zz", { "ale", "apple", "monkey", "plea" }, "" },
    };
    for (Test& t : tests)
        t.run();
    return 0;
}

File: p524m/find_longest_word.py
from typing import List


class Solution:
    def findLongestWordV1(self, s: str, d: List[str]) -> str:
        ans = ""
        for word in d:
            if len(word) > len(ans) or (len(word) == len(ans) and word < ans):
                # match word in s
                i = 0
                for c in s:
                    if i < len(word) and word[i] == c:
                        i += 1
                if i == len(word):
                    ans = word
        return ans

    def findLongestWordV2(self, s: str, d: List[str]) -> str:
        ans = ""
        for w in d:
            if (-len(w), w) < (-len(ans), ans):
                it = iter(s)
                if all(c in it for c in w):
                    ans = w
        return ans


# TESTS
for s, d, expected in [
    ("abpcplea", ["ale", "apple", "monkey", "plea"], "apple"),
    ("abpcplea", ["a", "b", "c"], "a"),
    ("zz", ["ale", "apple", "monkey", "plea"], ""),
]:
    sol = Solution()
    actual = sol.findLongestWordV1(s, d)
    print("The longest word in", d, "matching", s, "->", actual)
    assert actual == expected
    assert expected == sol.findLongestWordV2(s, d)
